File: services/notification_server/services/health_service.py
import asyncio
import logging
from datetime import datetime, timezone

from database.db_config import get_engine
from models.data_check import Check, CheckData, HealthReport
from communication.communication import test_connection, send_sample_message

logger = logging.getLogger(__name__)


def _run_async(coro):
    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            import concurrent.futures
            with concurrent.futures.ThreadPoolExecutor() as pool:
                return pool.submit(asyncio.run, coro).result()
        return loop.run_until_complete(coro)
    except Exception:
        return asyncio.run(coro)


def _build_check(name: str, is_up: bool, ready_label: str = "READY") -> Check:
    return Check(
        data=CheckData(
            from_=datetime.now(timezone.utc).isoformat(),
            status=ready_label if is_up else "DOWN",
        ),
        name=name,
        status="UP" if is_up else "DOWN",
    )


def _build_report(checks: list) -> HealthReport:
    overall = "UP" if all(c.status == "UP" for c in checks) else "DOWN"
    return HealthReport(status=overall, checks=checks)


def verify_ready() -> bool:
    try:
        with get_engine().connect():
            return True
    except Exception as exc:
        logger.warning("DB readiness check failed: %s", exc)
        return False


def verify_alive() -> bool:
    try:
        from sqlalchemy import text
        with get_engine().connect() as conn:
            conn.execute(text("SELECT 1"))
            return True
    except Exception as exc:
        logger.warning("DB liveness check failed: %s", exc)
        return False


def verify_nats_ready() -> bool:
    try:
        return _run_async(test_connection())
    except Exception:
        return False


def verify_nats_alive() -> bool:
    try:
        return _run_async(send_sample_message())
    except Exception:
        return False


def construct_ready_body() -> HealthReport:
    return _build_report([
        _build_check("Notification Service — Database readiness check", verify_ready()),
        _build_check("Notification Service — NATS readiness check",     verify_nats_ready()),
    ])


def construct_alive_body() -> HealthReport:
    return _build_report([
        _build_check("Notification Service — Database liveness check", verify_alive(),      "LIVE"),
        _build_check("Notification Service — NATS liveness check",     verify_nats_alive(), "LIVE"),
    ])

File: services/user_profile_server/services/healthService.go
package services

import (
	"time"
	"users_api/communication"
	"users_api/database"
	"users_api/models"
)

var StartTime time.Time

func Now() time.Time {
	return time.Now()
}

func VerifyHealth() models.GeneralCheck {
	dbOk := database.VerifyDatabaseConnection()
	natsOk := communication.ConnectToNATS().HealthCheckNATS()
	fromTime := time.Now()

	checks := []models.HealthCheck{
		buildCheck("User Profile Service — Database liveness check", dbOk, fromTime),
		buildCheck("User Profile Service — NATS liveness check", natsOk, fromTime),
	}

	return buildReport(checks)
}

func VerifyReadyHealth() models.GeneralCheck {
	dbOk := database.VerifyDatabaseReady()
	natsOk := communication.ConnectToNATS().ReadyNats()
	fromTime := time.Now()

	checks := []models.HealthCheck{
		buildCheck("User Profile Service — Database readiness check", dbOk, fromTime),
		buildCheck("User Profile Service — NATS readiness check", natsOk, fromTime),
	}

	return buildReport(checks)
}

func buildCheck(name string, isUp bool, from time.Time) models.HealthCheck {
	status := "DOWN"
	dataStatus := "DOWN"
	if isUp {
		status = "UP"
		dataStatus = "READY"
	}
	return models.HealthCheck{
		Data:   models.HealthData{From: from, Status: dataStatus},
		Name:   name,
		Status: status,
	}
}

func buildReport(checks []models.HealthCheck) models.GeneralCheck {
	overall := "UP"
	for _, c := range checks {
		if c.Status != "UP" {
			overall = "DOWN"
			break
		}
	}
	return models.GeneralCheck{
		Status:  overall,
		Checks:  checks,
		Version: "1.0.0",
		Uptime:  time.Since(StartTime).String(),
	}
}
